### PR TITLE
Microbatching fixes

### PR DESCRIPTION
**Pull request by dakinggg**
Multiple fixes:
(1) Hard error on dtms auto. We can support this later, but right now it crashes in a less obvious way.
(2) Adjust microbatch splitting logic so that microbatch sizes that do not evenly divide global batch size work.
(3) Adjust all the masked means in loss computation to first mean per sample and then mean across samples. This preserves the math for varying batch size. Let me know if you think this is wrong.

Adds a unit test for the new helper function added. We should add more comprehensive testing for the loss function, but deferring that to a separate PR.

Note: KL still differs slightly because logits themselves differ slightly between batch sizes. Accepting this for now.

Manual tests:
Loss is now the same with different mb sizes (1-4), and is the same as main branch with mb size 1. Previously the loss was quite different when using different mb sizes.

PPO:
Screenshot 2025-05-22 at 11 13 12 AM

GRPO:
Screenshot 2025-05-22 at 11 12 55 AM

### CODE REVIEW

**Review by gupta-abhay — APPROVED** [review]
this is fire !!! THANK YOU !!!

**Review by bcui-db — APPROVED** [review]
This is awesome thank you!!
